cirq/circuits/circuit.py

```diff
@@ -1387,6 +1389,48 @@ def tetris_concat(
 
         return cirq.Circuit(buffer[offset : offset + n_acc])
 
+    def get_qubit_factors(self) -> List[Set['cirq.Qid']]:
+        """
+        Divide circuit's qubits into qubit factors such that there are
+        no entangling gates between the qubits belonging to different factors.
+        If this is not possible, a sequence with a single factor (the whole set of
+        circuit's qubits) will be returned.
+        """
+        uf = networkx.utils.UnionFind(self.all_qubits())
+        for op in self.all_operations():
+            if len(op.qubits) > 1:
+                uf.union(*op.qubits)
+        return sorted(uf.to_sets(), key=min)
+
+    def factorize(self: CIRCUIT_TYPE) -> Iterable[CIRCUIT_TYPE]:
+        """Factorize circuit into a sequence of circuits, if it's possible (i.e. if
+        the circuit qubits can be divided into two or more groups of qubits
+        such that there are no entangling gates between them).
+        If this is not possible, will return the set consisting of the single
+        circuit (this one).
+        """
+
+        qubit_factors = self.get_qubit_factors()
+        if len(qubit_factors) == 1:
+            return (self,)
+        return (
+            self._with_sliced_moments([m[qubits] for m in self.moments]) for qubits in qubit_factors
+        )
+
+    def slice_by_qubit(
+        self: CIRCUIT_TYPE, with_qubits: Iterable['cirq.Qid']
```

Comment:
`Iterable['cirq.Qid']` is probably not the best type to use here, since there's no guarantee that an `Iterable` can be iterated over more than once; it could be a generator that gets exhausted the first time you iterate over it, then is empty if you try to iterate over it again. Something like `Collection['cirq.Qid']` would be better.

Reply:
Done.